dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -938,13 +938,14 @@ public Knowledge visitSameScope(Relation rel, String specificScope) {
                     Thread thread1 = e1.getThread();
                     Thread thread2 = e2.getThread();
                     if (specificScope != null) { // scope specified
-                        if (thread1.getScopeHierarchy().sameAtHigherScope(thread2.getScopeHierarchy(), specificScope)) {
+                        if (thread1.getScopeHierarchy().canSyncAtScope(thread2.getScopeHierarchy(), specificScope)) {
                             must.add(e1, e2);
                         }
                     } else {
                         String scope1 = Tag.getScopeTag(e1, program.getArch());
                         String scope2 = Tag.getScopeTag(e2, program.getArch());
-                        if (scope1.equals(scope2) && !scope1.isEmpty() && thread1.getScopeHierarchy().sameAtHigherScope(thread2.getScopeHierarchy(), scope1)) {
+                        if (!scope1.isEmpty() && !scope2.isEmpty() && thread1.getScopeHierarchy().canSyncAtScope(thread2.getScopeHierarchy(), scope1)
+                                && thread2.getScopeHierarchy().canSyncAtScope(thread1.getScopeHierarchy(), scope2)) {
```

Comment:
Do we really need both checks? Isn't the check symmetric?

Reply:
I think so, because a morally strong pair requires that `each operation is strong and specifies a scope that includes the thread executing the other operation`. `thread1 \in scope2` doesn't guarantee `thread2 \in scope1`.

Reply:
Ok, I missed the fact that the second parameter is different, i.e. `scope1` vs `scope2`